File: packages/kindling-core/src/types/pin.ts
/**
 * Pin types and definitions
 *
 * A Pin marks an observation or summary as important
 * (user-curated for retrieval prioritization)
 */

import type { ID, Timestamp, ScopeIds } from './common.js';

/**
 * Type of entity that can be pinned
 */
export type PinTargetType = 'observation' | 'summary';

/**
 * Pin entity
 *
 * Immutable - delete to remove
 */
export interface Pin {
  /** Unique identifier (UUIDv4) */
  id: ID;

  /** Type of entity being pinned */
  targetType: PinTargetType;

  /** ID of the pinned entity (observation or summary) */
  targetId: ID;

  /**
   * Optional user-provided explanation for why this is pinned
   * e.g., "Critical context for auth flow"
   */
  reason?: string;

  /** Timestamp when pin was created (epoch milliseconds) */
  createdAt: Timestamp;

  /**
   * Optional expiration timestamp (epoch milliseconds)
   * If set, pin is only active while expiresAt > now
   * Supports time-bound pins (e.g., session-only)
   */
  expiresAt?: Timestamp;

  /** Isolation dimensions for scoped queries */
  scopeIds: ScopeIds;
}

/**
 * Input for creating a new pin
 * Makes id and createdAt optional (will be auto-generated)
 */
export interface PinInput {
  id?: ID;
  targetType: PinTargetType;
  targetId: ID;
  reason?: string;
  createdAt?: Timestamp;
  expiresAt?: Timestamp;
  scopeIds: ScopeIds;
}

/**
 * All valid pin target types
 */
export const PIN_TARGET_TYPES: readonly PinTargetType[] = ['observation', 'summary'] as const;

/**
 * Type guard to check if a string is a valid PinTargetType
 */
export function isPinTargetType(value: unknown): value is PinTargetType {
  return typeof value === 'string' && PIN_TARGET_TYPES.includes(value as PinTargetType);
}

/**
 * Check if a pin is active (not expired) at a given timestamp
 */
export function isPinActive(pin: Pin, now: Timestamp): boolean {
  return pin.expiresAt === undefined || pin.expiresAt > now;
}

File: crates/kindling-types/src/pin.rs
use serde::{Deserialize, Serialize};

#[cfg(feature = "ts-rs")]
use ts_rs::TS;

use crate::common::{Id, ScopeIds, Timestamp};

/// Type of entity that can be pinned.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-rs", derive(TS), ts(export, export_to = "../bindings/"))]
#[serde(rename_all = "snake_case")]
pub enum PinTargetType {
    Observation,
    Summary,
}

impl PinTargetType {
    pub const ALL: &'static [PinTargetType] = &[PinTargetType::Observation, PinTargetType::Summary];
}

/// Pinned reference to an observation or summary.
///
/// Mirrors `Pin` in `packages/kindling-core/src/types/pin.ts`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-rs", derive(TS), ts(export, export_to = "../bindings/"))]
#[serde(rename_all = "camelCase")]
pub struct Pin {
    pub id: Id,
    pub target_type: PinTargetType,
    pub target_id: Id,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional))]
    pub reason: Option<String>,
    #[cfg_attr(feature = "ts-rs", ts(type = "number"))]
    pub created_at: Timestamp,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional, type = "number"))]
    pub expires_at: Option<Timestamp>,
    pub scope_ids: ScopeIds,
}

/// Input for creating a new pin. Optional fields are auto-generated.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "ts-rs", derive(TS), ts(export, export_to = "../bindings/"))]
#[serde(rename_all = "camelCase")]
pub struct PinInput {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional))]
    pub id: Option<Id>,
    pub target_type: PinTargetType,
    pub target_id: Id,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional))]
    pub reason: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional, type = "number"))]
    pub created_at: Option<Timestamp>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[cfg_attr(feature = "ts-rs", ts(optional, type = "number"))]
    pub expires_at: Option<Timestamp>,
    pub scope_ids: ScopeIds,
}

/// True iff the pin has not expired at `now` (epoch ms).
pub fn is_pin_active(pin: &Pin, now: Timestamp) -> bool {
    match pin.expires_at {
        None => true,
        Some(exp) => exp > now,
    }
}
